File: cryptography-hash/src/test/java/dev/retrotv/crypto/hash/HashTest.java
package dev.retrotv.crypto.hash;

import dev.retrotv.crypto.exception.AlgorithmNotFoundException;
import dev.retrotv.crypto.hash.enums.EHash;
import dev.retrotv.crypto.util.Base64CodecUtils;
import dev.retrotv.crypto.util.HEXCodecUtils;
import dev.retrotv.data.enums.EncodeFormat;
import org.json.JSONObject;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.RepeatedTest;
import org.junit.jupiter.api.Test;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.net.URL;

import static org.junit.jupiter.api.Assertions.*;

class HashTest {
    private final String password = "The quick brown fox jumps over the lazy dog";
    private final URL checksum = this.getClass().getClassLoader().getResource("hash_code");
    private final URL resource = this.getClass().getClassLoader().getResource("hash_code_test_file.txt");

    @RepeatedTest(value = 100, name = "{displayName} {currentRepetition}/{totalRepetitions}")
    @DisplayName("CRC-32 알고리즘으로 해싱")
    void test_crc32() {
        hashingTest(EHash.CRC32);
    }

    @RepeatedTest(value = 100, name = "{displayName} {currentRepetition}/{totalRepetitions}")
    @DisplayName("MD2 알고리즘으로 해싱")
    void test_md2() {
        hashingTest(EHash.MD2);
    }

    @RepeatedTest(value = 100, name = "{displayName} {currentRepetition}/{totalRepetitions}")
    @DisplayName("MD5 알고리즘으로 해싱")
    void test_md5() {
        hashingTest(EHash.MD5);
    }

    @RepeatedTest(value = 100, name = "{displayName} {currentRepetition}/{totalRepetitions}")
    @DisplayName("SHA-1 알고리즘으로 해싱")
    void test_sha1() {
        hashingTest(EHash.SHA1);
    }

    @RepeatedTest(value = 100, name = "{displayName} {currentRepetition}/{totalRepetitions}")
    @DisplayName("SHA-224 알고리즘으로 해싱")
    void test_sha224() {
        hashingTest(EHash.SHA224);
    }

    @RepeatedTest(value = 100, name = "{displayName} {currentRepetition}/{totalRepetitions}")
    @DisplayName("SHA-256 알고리즘으로 해싱")
    void test_sha256() {
        hashingTest(EHash.SHA256);
    }

    @RepeatedTest(value = 100, name = "{displayName} {currentRepetition}/{totalRepetitions}")
    @DisplayName("SHA-384 알고리즘으로 해싱")
    void test_sha384() {
        hashingTest(EHash.SHA384);
    }

    @RepeatedTest(value = 100, name = "{displayName} {currentRepetition}/{totalRepetitions}")
    @DisplayName("SHA-512 알고리즘으로 해싱")
    void test_sha512() {
        hashingTest(EHash.SHA512);
    }

    @RepeatedTest(value = 100, name = "{displayName} {currentRepetition}/{totalRepetitions}")
    @DisplayName("SHA-512/224 알고리즘으로 해싱")
    void test_sha512224() {
        hashingTest(EHash.SHA512224);
    }

    @RepeatedTest(value = 100, name = "{displayName} {currentRepetition}/{totalRepetitions}")
    @DisplayName("SHA-512/256 알고리즘으로 해싱")
    void test_sha512256() {
        hashingTest(EHash.SHA512256);
    }

    @RepeatedTest(value = 100, name = "{displayName} {currentRepetition}/{totalRepetitions}")
    @DisplayName("SHA3-224 알고리즘으로 해싱")
    void test_sha3224() {
        hashingTest(EHash.SHA3224);
    }

    @RepeatedTest(value = 100, name = "{displayName} {currentRepetition}/{totalRepetitions}")
    @DisplayName("SHA3-256 알고리즘으로 해싱")
    void test_sha3256() {
        hashingTest(EHash.SHA3256);
    }

    @RepeatedTest(value = 100, name = "{displayName} {currentRepetition}/{totalRepetitions}")
    @DisplayName("SHA3-384 알고리즘으로 해싱")
    void test_sha3384() {
        hashingTest(EHash.SHA3384);
    }

    @RepeatedTest(value = 100, name = "{displayName} {currentRepetition}/{totalRepetitions}")
    @DisplayName("SHA3-512 알고리즘으로 해싱")
    void test_sha3512() {
        hashingTest(EHash.SHA3512);
    }

    @Test
    @DisplayName("getInstance(String) 테스트")
    void test_getInstance() {
        Hash h1 = Hash.getInstance("MD5");
        Hash h2 = Hash.getInstance(EHash.MD5);

        assertEquals(HEXCodecUtils.encode(h1.hashing(password)), HEXCodecUtils.encode(h2.hashing(password)));
        assertEquals(Base64CodecUtils.encode(h1.hashing(password)), Base64CodecUtils.encode(h2.hashing(password)));
    }

    @Test
    @DisplayName("AlgorithmNotFoundException 테스트")
    void test_algorithmNotFoundException() {
        assertThrows(AlgorithmNotFoundException.class, () ->
            Hash.getInstance("WRONG_ALGORITHM")
        );
    }

    private void hashingTest(EHash algorithm) {
        passwordHashTest(algorithm);
        fileHashTest(algorithm);
    }

    private void passwordHashTest(EHash algorithm) {
        Hash h = Hash.getInstance(algorithm);
        assertTrue(h.matches(password.getBytes(), getHash(algorithm)));
        assertEquals(HEXCodecUtils.encode(h.hashing(password)), getHash(algorithm));
        assertEquals(HEXCodecUtils.encode(h.hashing(password, java.nio.charset.StandardCharsets.UTF_8)), getHash(algorithm));
    }

    private void fileHashTest(EHash algorithm) {
        BinaryHash h = (BinaryHash) Hash.getInstance(algorithm);

        assertTrue(
                h.matches(
                        getResourceFileBytes(resource),
                        HEXCodecUtils.encode(h.hashing(getResourceFileBytes(resource)))
                )
        );

        assertFalse(h.matches(getResourceFileBytes(resource), (byte[]) null));
        assertFalse(h.matches(getResourceFileBytes(resource), (String) null, EncodeFormat.HEX));
    }

    private String getHash(EHash algorithm) {
        try {
            JSONObject jsonObject = new JSONObject(readJson());
            JSONObject file = jsonObject.getJSONObject("hash_code_test_file");
            switch (algorithm) {
                case CRC32: return file.getString(EHash.CRC32.label());
                case MD2: return file.getString(EHash.MD2.label());
                case MD5: return file.getString(EHash.MD5.label());
                case SHA1: return file.getString(EHash.SHA1.label());
                case SHA224: return file.getString(EHash.SHA224.label());
                case SHA256: return file.getString(EHash.SHA256.label());
                case SHA384: return file.getString(EHash.SHA384.label());
                case SHA512: return file.getString(EHash.SHA512.label());
                case SHA512224: return file.getString(EHash.SHA512224.label());
                case SHA512256: return file.getString(EHash.SHA512256.label());
                case SHA3224: return file.getString(EHash.SHA3224.label());
                case SHA3256: return file.getString(EHash.SHA3256.label());
                case SHA3384: return file.getString(EHash.SHA3384.label());
                case SHA3512: return file.getString(EHash.SHA3512.label());
                default: return null;
            }
        } catch (IOException e) {
            return null;
        }
    }

    private String readJson() throws IOException {
        StringBuilder sb = new StringBuilder();
        try (BufferedReader reader = new BufferedReader(new FileReader(checksum.getFile()))) {
            String line;
            while ((line = reader.readLine()) != null) {
                sb.append(line).append("\n");
            }
        }
        return sb.toString();
    }

    private byte[] getResourceFileBytes(URL resourceUrl) {
        if (resourceUrl == null) return new byte[0];
        try (BufferedReader reader = new BufferedReader(new FileReader(resourceUrl.getFile()))) {
            StringBuilder sb = new StringBuilder();
            String line;
            while ((line = reader.readLine()) != null) {
                sb.append(line).append("\n");
            }
            return sb.toString().getBytes();
        } catch (IOException e) {
            return new byte[0];
        }
    }
}

File: cryptography-hash/src/test/kotlin/dev/retrotv/crypto/hash/HashTest.kt
package dev.retrotv.crypto.hash

import dev.retrotv.crypto.enums.EHash
import dev.retrotv.crypto.enums.EHash.*
import dev.retrotv.crypto.exception.AlgorithmNotFoundException
import dev.retrotv.crypto.util.CodecUtils
import dev.retrotv.data.enums.EncodeFormat

import org.json.JSONObject
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.RepeatedTest
import org.junit.jupiter.api.Test
import java.io.BufferedReader
import java.io.FileReader
import java.io.IOException

class HashTest {
    private val password = "The quick brown fox jumps over the lazy dog"
    private val checksum = this.javaClass.getClassLoader().getResource("hashing_code")
    private val resource = this.javaClass.getClassLoader().getResource("hashing_code_test_file.txt")

    @RepeatedTest(100, name = "{displayName} {currentRepetition}/{totalRepetitions}")
    @DisplayName("CRC-32 알고리즘으로 해싱")
    fun test_crc32() {
        hashingTest(CRC32)
    }

    @RepeatedTest(100, name = "{displayName} {currentRepetition}/{totalRepetitions}")
    @DisplayName("MD2 알고리즘으로 해싱")
    fun test_md2() {
        hashingTest(MD2)
    }

    @RepeatedTest(100, name = "{displayName} {currentRepetition}/{totalRepetitions}")
    @DisplayName("MD5 알고리즘으로 해싱")
    fun test_md5() {
        hashingTest(MD5)
    }

    @RepeatedTest(100, name = "{displayName} {currentRepetition}/{totalRepetitions}")
    @DisplayName("SHA-1 알고리즘으로 해싱")
    fun test_sha1() {
        hashingTest(SHA1)
    }

    @RepeatedTest(100, name = "{displayName} {currentRepetition}/{totalRepetitions}")
    @DisplayName("SHA-224 알고리즘으로 해싱")
    fun test_sha224() {
        hashingTest(SHA224)
    }

    @RepeatedTest(100, name = "{displayName} {currentRepetition}/{totalRepetitions}")
    @DisplayName("SHA-256 알고리즘으로 해싱")
    fun test_sha256() {
        hashingTest(SHA256)
    }

    @RepeatedTest(100, name = "{displayName} {currentRepetition}/{totalRepetitions}")
    @DisplayName("SHA-384 알고리즘으로 해싱")
    fun test_sha384() {
        hashingTest(SHA384)
    }

    @RepeatedTest(100, name = "{displayName} {currentRepetition}/{totalRepetitions}")
    @DisplayName("SHA-512 알고리즘으로 해싱")
    fun test_sha512() {
        hashingTest(SHA512)
    }

    @RepeatedTest(100, name = "{displayName} {currentRepetition}/{totalRepetitions}")
    @DisplayName("SHA-512/224 알고리즘으로 해싱")
    fun test_sha512224() {
        hashingTest(SHA512224)
    }

    @RepeatedTest(100, name = "{displayName} {currentRepetition}/{totalRepetitions}")
    @DisplayName("SHA-512/256 알고리즘으로 해싱")
    fun test_sha512256() {
        hashingTest(SHA512256)
    }

    @RepeatedTest(100, name = "{displayName} {currentRepetition}/{totalRepetitions}")
    @DisplayName("SHA3-224 알고리즘으로 해싱")
    fun test_sha3224() {
        hashingTest(SHA3224)
    }

    @RepeatedTest(100, name = "{displayName} {currentRepetition}/{totalRepetitions}")
    @DisplayName("SHA3-256 알고리즘으로 해싱")
    fun test_sha3256() {
        hashingTest(SHA3256)
    }

    @RepeatedTest(100, name = "{displayName} {currentRepetition}/{totalRepetitions}")
    @DisplayName("SHA3-384 알고리즘으로 해싱")
    fun test_sha3384() {
        hashingTest(SHA3384)
    }

    @RepeatedTest(100, name = "{displayName} {currentRepetition}/{totalRepetitions}")
    @DisplayName("SHA3-512 알고리즘으로 해싱")
    fun test_sha3512() {
        hashingTest(SHA3512)
    }

    @Test
    @DisplayName("getInstance(String) 테스트")
    fun test_getInstance() {
        val h1 = Hash.getInstance("MD5")
        val h2 = Hash.getInstance(MD5)

        assertEquals(CodecUtils.encode(h1.hashing(this.password)), CodecUtils.encode(h2.hashing(this.password)))
    }

    @Test
    @DisplayName("AlgorithmNotFoundException 테스트")
    fun test_algorithmNotFoundException() {
        assertThrows(AlgorithmNotFoundException::class.java) {
            Hash.getInstance("WRONG_ALGORITHM")
        }
    }

    private fun hashingTest(algorithm: EHash) {
        passwordHashTest(algorithm)
        fileHashTest(algorithm)
    }

    private fun passwordHashTest(algorithm: EHash) {
        val h = Hash.getInstance(algorithm)
        assertTrue(h.matches(password.toByteArray(), getHash(algorithm)))
        assertEquals(CodecUtils.encode(h.hashing(password), EncodeFormat.HEX), getHash(algorithm))
        assertEquals(CodecUtils.encode(h.hashing(password, Charsets.UTF_8), EncodeFormat.HEX), getHash(algorithm))
    }

    private fun fileHashTest(algorithm: EHash) {
        val h: BinaryHash = Hash.getInstance(algorithm)
        assertTrue(
            h.matches(
                resource?.file!!.toByteArray(),
                CodecUtils.encode(h.hashing(resource.file!!.toByteArray()))
            )
        )

        assertFalse(h.matches(resource.file!!.toByteArray(), null as ByteArray?))
        assertFalse(h.matches(resource.file!!.toByteArray(), null as String?, EncodeFormat.HEX))
    }

    @Throws(IOException::class)
    private fun getHash(algorithm: EHash): String? {
        val jsonObject = JSONObject(readJson())
        val file = jsonObject.getJSONObject("hashing_code_test_file")
        return when (algorithm) {
            CRC32 -> file.getString(CRC32.label())
            MD2 -> file.getString(MD2.label())
            MD5 -> file.getString(MD5.label())
            SHA1 -> file.getString(SHA1.label())
            SHA224 -> file.getString(SHA224.label())
            SHA256 -> file.getString(SHA256.label())
            SHA384 -> file.getString(SHA384.label())
            SHA512 -> file.getString(SHA512.label())
            SHA512224 -> file.getString(SHA512224.label())
            SHA512256 -> file.getString(SHA512256.label())
            SHA3224 -> file.getString(SHA3224.label())
            SHA3256 -> file.getString(SHA3256.label())
            SHA3384 -> file.getString(SHA3384.label())
            SHA3512 -> file.getString(SHA3512.label())
        }
    }

    @Throws(IOException::class)
    private fun readJson(): String {
        if (checksum == null) {
            throw IOException()
        }

        var json: String
        BufferedReader(FileReader(checksum.file)).use { reader ->
            val sb = StringBuilder()
            var line = reader.readLine()
            while (line != null) {
                sb.append(line)
                sb.append("\n")
                line = reader.readLine()
            }
            json = sb.toString()
        }

        return json
    }
}